x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/create_security_rule_type_wrapper.ts

```diff
@@ -530,7 +530,12 @@ export const createSecurityRuleTypeWrapper: CreateSecurityRuleTypeWrapper =
                     enrichmentDurations: result.enrichmentTimes,
                   },
                 });
-              } else if (wroteWarningStatus && !hasError && !result.warning) {
+              } else if (
+                !isEmpty(warningMessage) &&
+                wroteWarningStatus &&
```

Comment:
This block should not depend on `wroteWarningStatus` - that flag was used to keep track of whether a rule warning has **already** been written to the rule execution status in order to avoid writing a "success" status afterwards that would hide the warning. However, now it can have multiple meanings - it's set to true when warnings are written, but also set to true in the `disabledActions` logic when we haven't written warnings but need to write them later.

I think what we need to do here is separate the "warning writing" logic from the "metric writing". Since https://github.com/elastic/kibana/pull/167551, we write warnings at the beginning of rule execution then store the warning message and re-write the same warning at the end of rule execution with metrics. Instead we should make it possible through the rule execution logger to write metrics independently of the status so we don't have to store and re-write the warnings. If we do that I don't think we'll need this extra block here to re-write warnings that were written earlier.

Reply:
We should aim for a design where we don't have to keep track of warning messages or manually combine them - we simply write the warning when it's first detected.